server/game/effects.js

```diff
@@ -1451,31 +1451,47 @@ const Effects = {
         return {
             targetType: 'player',
             apply: function(player, context) {
-                const shadows = player.shadows;
-                const revealFunc = reveal => shadows.includes(reveal);
+                // making a snapshot of the shadows area through a shallow copy. This way eventual changes to
+                // player.shadows array content (first level) are not reflected in the snapshot
+                const shadows = [...player.shadows];
 
+                // we define the reveal function just once, so it must be bound to the player to keep track of the
+                // content of the shadows area when the function is called by the engine
+                const revealFunc = reveal => player.shadows.includes(reveal);
+                context.game.cardVisibility.addRule(revealFunc);
+
+                // saving in the context a reference to the revealFunc (so we can remove it from the engine when the
+                // effect expires) and the actual snapshot of the cards in the shadows area for which we are resolving
+                // the reveal game action (to avoid a second resolution of the same game action in case of reapply)
                 context.revealShadows = context.revealShadows || {};
                 context.revealShadows[player.name] = {
                     revealFunc,
                     revealed: shadows
                 };
-                context.game.cardVisibility.addRule(revealFunc);
 
-                context.game.resolveGameAction(GameActions.revealCards({
-                    cards: shadows,
-                    player,
-                    revealWithMessage: false,
-                    highlight: false,
-                    source: context.source
-                }), context);
+                // resolving the 'reveal' game action for cards in shadows (if there are any)
+                if(shadows.length > 0) {
```

Comment:
This `if` is actually unnecessary, since the `allow` method for `RevealCard` action returns false for empty arrays. But it's consistent with the rest of this effect, so 🤷‍♀️

Reply:
Completely forgot that the `resolveGameAction` method passes through `action.allow(props)` before actually resolving, good catch! I might push a tweak through to remove it *at some point* :P